apps/dashboard/test/models/hpc_module_test.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class HpcModuleTest < ActiveSupport::TestCase
+  test 'all safely reads from inaccessabile directories' do
+    with_modified_env({ OOD_MODULE_FILE_DIR: '/dev/null' }) do
+      assert_equal [], HpcModule.all('owens')
+    end
+  end
+
+  test 'all safely reads invalid json' do
+    Dir.mktmpdir do |dir|
+      with_modified_env({ OOD_MODULE_FILE_DIR: dir }) do
+        `echo '{this is bad json}' > #{dir}/owens.json`
+        assert_equal [], HpcModule.all('owens')
+      end
+    end
+  end
+
+  test 'reads a simple file' do
+    with_modified_env({ OOD_MODULE_FILE_DIR: "#{Rails.root}/test/fixtures/modules/" }) do
+      # NOTE: that there are no duplicates and rstudio has no version
+      assert_equal(['jupyter/1', 'jupyter/2', 'rstudio'], HpcModule.all('simple').map(&:to_s))
+    end
+  end
+
+  test 'default version' do
+    m = HpcModule.new('test')
+    assert m.default?
+    assert m.version.nil?
+  end
+
+  test 'module with version version' do
+    m = HpcModule.new('test', version: 9001)
+    assert !m.default?
+    assert_equal m.version, '9001' # we gave an int, got back a string
+  end
+
+  test 'module equivalence' do
+    m1 = HpcModule.new('test', version: 9001)
+    m2 = HpcModule.new('test', version: 9001)
+
+    assert m1 == m2
+    assert m1.eql?(m2)
```

Comment:
Not sure if this "eql?" is the alias from hpc_module.rb or if it the actual ruby "eql?".

Without understanding the why, my thoughts are we shouldn't override out of the box ruby functions, as it can cause lots of confusion.  If there is a good reason why we should do this type of thing, let's discuss please and understand the benefits.  If we absolutely have to override, then we should comment the reasoning in the code.

Reply:
> If there is a good reason why we should do this type of thing, let's discuss please and understand the benefits.

This is exactly the time and place to discus, and it's not unreasonable to ask why.

The real feature I'm looking for is something like this.  Another reason to rede
```ruby
if m_module == "rstudio/123"
  # do something
else
  # do something else
end
```

Lastly, it's about uniqueness. Let's remember what these objects represent - modules like `rstudio/3`. It is unique based on that string of name + "/" + version. 

This is the behavior I'm looking to get. They are equal by value but not by identity.
```irb
2.7.1 :001 > "a".eql?("a")
 => true 
2.7.1 :002 > "a" == "a"
 => true 
2.7.1 :003 > "a".object_id == "a".object_id
 => false 
```

Reply:
> we should comment the reasoning in the code.

Comment added.

Reply:
You're right - we don't need the additional complexity.